include/qore/StreamReader.h

```diff
@@ -0,0 +1,363 @@
+         // Won't overflow because the buffer actually has real capacity of bufCapacity+1.
+         buf[bufSize] = '\0';
+
+         qore_size_t eolLen;
+         const char* p = findEolInBuffer(*eolstr, eolLen, rc==0);
+         assert(eolLen >= 0);
+
+         if (p) { // Found end of line.
+            assert(p >= buf);
+            qore_size_t dataSize = p - buf;
+            line->concat(buf, dataSize + (trim ? 0 : eolLen));
+            // Move remaining data from middle to the front of the buffer.
+            bufSize -= dataSize + eolLen;
+            memmove(buf, p + eolLen, bufSize);
+            return line.release();
+         }
+         else {
+            if (rc > 0) {
+               qore_size_t dataSize = (bufSize >= eolLen) ? bufSize - eolLen : 0;
+               line->concat(buf, dataSize);
+               // Move remaining data from middle to the front of the buffer.
+               assert(bufSize >= dataSize);
+               bufSize -= dataSize;
+               memmove(buf, buf + dataSize, bufSize);
+               continue;
+            }
+            else if (rc == 0) { // At the end of stream.
+               assert(bufCapacity != bufSize);
+               line->concat(buf, bufSize);
+               bufSize = 0;
+               return line.release();
+            }
+         }
+      }
+   }
+
+   //! Read binary data from the stream.
+   /** @param limit max amount of data to read; if equal to -1, all data will be read, if equal to 0, no data will be read
+    */
+   DLLLOCAL BinaryNode* readBinary(int64 limit, ExceptionSink* xsink) {
+      SimpleRefHolder<BinaryNode> b(new BinaryNode());
+      if (limit >= 0 && static_cast<int64>(bufSize) >= limit) {
+         b->append(buf, limit);
+         shiftBuffer(limit);
+         return b->empty() ? 0 : b.release();
+      }
+      else { // Either limit is -1 or we need more data than is currently in the buffer.
+         int64 toRead = limit;
+         b->append(buf, bufSize);
+         toRead -= bufSize;
+         bufSize = 0;
+         while (true) {
+            qore_size_t bytes = limit < 0 ? bufCapacity : QORE_MIN(static_cast<qore_size_t>(toRead), bufCapacity);
+            int64 rc = fillBuffer(bytes, xsink);
+            if (*xsink)
+               return 0;
+            if (rc == 0) // Input stream end.
+               return b->empty() ? 0 : b.release();
+
+            b->append(buf, bufSize);
+            toRead -= bufSize;
+            bufSize = 0;
+            if (toRead == 0)
+               return b.release();
+         }
+      }
+   }
+
+   DLLLOCAL int64 readi1(ExceptionSink* xsink) {
+      char i = 0;
+      if (!prepareEnoughData(1, xsink))
+         return 0;
+      i = buf[0];
+      shiftBuffer(1);
+      return i;
+   }
+
+   DLLLOCAL int64 readi2(ExceptionSink* xsink) {
+      short i = 0;
+      if (!prepareEnoughData(2, xsink))
+         return 0;
+      i = reinterpret_cast<short*>(buf)[0];
+      shiftBuffer(2);
+      i = ntohs(i);
```

Comment:
but in our case the source data are in host order and we want them in big-endian (network) order, so we want host to network -> htons